### PR TITLE
ft. modernise py-libp2p

### PR DESCRIPTION
**Pull request by arcinston**
## Outdated PyProject.toml , revamp typechecks , linting and formating

related to Issue https://github.com/libp2p/py-libp2p/discussions/610

1. Updated pyproject.toml (and replace setup.py)

2. Use Ruff as linter and formatter (to replace black , isort ,pydocstyle , pycodestyle,pyflakes)
this has also made the pre-commit hook insanely faster
image

3. Add PyreFly as a better , faster & stricted static typechecker in addition to mypy
p.s. - Pyrefly has already helped find 600+ type errors (major & minor) which i have been working on reducing to 0

4. Updating the PreCommit hooks to
     a) use Ruff for linting & formatting 
     b) add PyRight for static type check in addition to mypy

### To-Do

- [x] Fixing all the newly detected static typing issues which have existed in the codebase
- [x] Updating the PreCommit Hooks for PyRight & Ruff
- [x] Updating all Github workflows to current spec
- [x] Final Checks and Testing

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by arcinston** [comment]
examples of issues found by pyright

image
in test_examples.py -> a float value is wrongly passed to the constructor of the gossipsub which expects an int

image
these functions which do not accept None as an argument (part of gossipsub.py)

**Comment by varun-r-mallya** [comment]
@arcinston can I join this effort ? Do I continue working on this by forking your branch or should I be doing something else ?

**Comment by arcinston** [comment]
> @arcinston can I join this effort ? Do I continue working on this by forking your branch or should I be doing something else ?


opening a pr to my branch works

**Comment by Aldorax** [comment]
> # examples of issues found by pyright
> image
> ## in test_examples.py -> a float value is wrongly passed to the constructor of the gossipsub which expects an int
> image
> ## these functions which do not accept None as an argument (part of gossipsub.py)

Hey there @arcinston I will be joining you with the PR. I will be making a PR to your branch soon with some fixes. Currently running tests.

**Comment by varun-r-mallya** [comment]
https://pyrefly.org/ is something that Meta released very recently and I think this would be a great addition to our type checking efforts as well.
https://github.com/facebook/pyrefly

**Comment by sukhman-sukh** [comment]
> > @arcinston can I join this effort ? Do I continue working on this by forking your branch or should I be doing something else ?
> 
> opening a pr to my branch works

I can't find your branch in the repo. How can I contribute to this PR?

**Comment by arcinston** [comment]
> https://pyrefly.org/ is something that Meta released very recently and I think this would be a great addition to our type checking efforts as well.
> https://github.com/facebook/pyrefly

this seems awesome, checking this out

**Comment by arcinston** [comment]
@varun-r-mallya pyrefly seems good

as we are using ruff now , the next best solution to me seems https://github.com/astral-sh/ty which is currently in beta
will give ruff + ty combo shoould be very effecient

in the mean while we can shift to pyright as it is still widely available and both ty and pyrefly are both in their early days

**Comment by seetadev** [comment]
@arcinston : This is coming along very nicely. Great work. 

Also, appreciate the contributions by @varun-r-mallya , @kaneki003 @sukhman-sukh and @Aldorax . Keep up the momentum everyone :) 

@arcinston : Will discuss in details and contribute towards arriving at a good conclusion on the issues related to factories.py and multiaddrs. Awesome progress.

CCing @pacrob , who can start review and share feedback and pointers as we arrive at a mid-point milestone in this pull request. His pointers are invaluable :)

Looking forward to py-libp2p becoming better and typesafe with the new stack :)

**Comment by acul71** [comment]
@seetadev 
Typechecks, linting issues for py-multiaddr fixed here:
https://github.com/acul71/py-multiaddr-fork/tree/fix/typecheck
The repository is archived. 
Unsure how to proceed.


# Code and Test Changes Documentation

This document outlines the changes made to the code and test files between the current version and commit e01dbd38f2c0464c0f78b556691d655265018cce.

## Overview of Changes

The main changes across the codebase include:
1. Enhanced type safety and protocol handling
   - Added comprehensive type hints throughout the codebase
   - Improved protocol validation and error handling
   - Enhanced type checking for protocol operations
2. Improved CID (Content Identifier) support
   - Added support for CIDv1 format
   - Enhanced CID conversion handling
   - Improved CID validation and parsing
3. Better sequence protocol implementation
   - Added proper sequence protocol support
   - Enhanced indexing and slicing operations
   - Improved sequence validation
4. More robust error handling
   - Added specific exception types
   - Enhanced error messages
   - Improved error recovery
5. Codec improvements and optimizations
   - Enhanced codec validation
   - Improved codec performance
   - Added new codec features
6. Enhanced string parsing and validation
   - Improved string format validation
   - Enhanced parsing error handling
   - Added support for new string formats
7. Improved protocol registry handling
   - Enhanced protocol registration
   - Improved protocol lookup
   - Added protocol validation
8. Comprehensive test coverage improvements
   - Added new test cases
   - Enhanced existing tests
   - Improved test organization

## Detailed Changes

### 1. Core Module Changes

#### multiaddr/multiaddr.py
The Multiaddr class has been significantly enhanced with the following changes:

1. Sequence Protocol Implementation
   - Added proper implementation of Python's sequence protocol
   - Enhanced support for indexing and slicing operations
   - Improved handling of negative indices
   - Added support for sequence operations (len, contains, etc.)

2. Type Hints and Annotations
   - Added comprehensive type hints throughout the class
   - Enhanced type safety for protocol operations
   - Improved type checking for method parameters and return values
   - Added support for generic types where appropriate

3. View Handling
   - Improved handling of keys, values, and items views
   - Enhanced view iteration and access
   - Added support for view operations
   - Improved view performance

4. Negative Indexing
   - Added support for negative indices in sequence operations
   - Enhanced index validation
   - Improved error handling for invalid indices
   - Added bounds checking for indices

5. Protocol Alias Handling
   - Improved handling of protocol aliases
   - Enhanced alias validation
   - Added support for alias operations
   - Improved alias lookup performance

6. Error Handling
   - Enhanced error handling for protocol lookups
   - Improved error messages
   - Added specific exception types
   - Enhanced error recovery

7. String Representation
   - Improved string representation handling
   - Enhanced string formatting
   - Added support for custom string formats
   - Improved string validation

Key changes:
```python
def __getitem__(self, idx: Union[int, slice]) -> Union[Any, Sequence[Any]]:
    if isinstance(idx, slice):
        return list(self)[idx]
    if idx < 0:
        idx = len(self)+idx
    for idx2, value in enumerate(self):
        if idx2 == idx:
            return value
    raise IndexError("Protocol value list index out of range")

def __contains__(self, value: object) -> bool:
    return collections.abc.Sequence.__contains__(self, value)
```

#### multiaddr/protocols.py
The protocols module has been enhanced with the following changes:

1. Protocol Registry Handling
   - Improved protocol registration process
   - Enhanced registry validation
   - Added support for registry operations
   - Improved registry performance

2. Type Safety
   - Enhanced type safety for protocol objects
   - Improved type checking
   - Added support for type validation
   - Enhanced type hints

3. Protocol Alias Handling
   - Improved handling of protocol aliases
   - Enhanced alias validation
   - Added support for alias operations
   - Improved alias lookup

4. Protocol Locking
   - Added support for protocol locking
   - Enhanced lock validation
   - Improved lock operations
   - Added lock safety checks

5. Error Handling
   - Enhanced error handling for protocol lookups
   - Improved error messages
   - Added specific exception types
   - Enhanced error recovery

6. String Parsing
   - Improved string parsing and validation
   - Enhanced parsing error handling
   - Added support for new string formats
   - Improved parsing performance

7. Protocol Validation
   - Added support for protocol validation
   - Enhanced validation rules
   - Improved validation performance
   - Added validation error handling

Key changes:
```python
def find_by_name(self, name: str) -> Protocol:
    try:
        return self._names_to_protocols[name]
    except KeyError:
        raise exceptions.ProtocolNotFoundError(name)

def find_by_code(self, code: int) -> Protocol:
    try:
        return self._codes_to_protocols[code]
    except KeyError:
        raise exceptions.ProtocolNotFoundError(code)
```

#### multiaddr/transforms.py
The transforms module has been enhanced with the following changes:

1. CID Conversion
   - Improved CID conversion handling
   - Enhanced conversion validation
   - Added support for new CID formats
   - Improved conversion performance

2. Error Handling
   - Enhanced error handling
   - Improved error messages
   - Added specific exception types
   - Enhanced error recovery

3. Type Safety
   - Enhanced type safety
   - Improved type checking
   - Added support for type validation
   - Enhanced type hints

4. CIDv1 Support
   - Added support for CIDv1 format
   - Enhanced CIDv1 validation
   - Improved CIDv1 handling
   - Added CIDv1 conversion

5. String Parsing
   - Enhanced string parsing and validation
   - Improved parsing error handling
   - Added support for new string formats
   - Improved parsing performance

6. Protocol Validation
   - Added support for protocol validation
   - Enhanced validation rules
   - Improved validation performance
   - Added validation error handling

7. Path-based Protocols
   - Added support for path-based protocols
   - Enhanced path validation
   - Improved path handling
   - Added path operations

Key changes:
```python
def string_iter(string: str) -> Generator[Tuple[Protocol, CodecBase, Optional[str]], None, None]:
    if not string:
        raise exceptions.StringParseError("Empty string", string)
    if not string.startswith('/'):
        raise exceptions.StringParseError("Must begin with /", string)
    # consume trailing slashes
    string = string.rstrip('/')
    sp = string.split('/')
    # skip the first element, since it starts with /
    sp.pop(0)
    while sp:
        element = sp.pop(0)
        if not element:  # Skip empty elements from multiple slashes
            continue
        try:
            proto = protocol_with_name(element)
            if proto.codec is None:
                codec = CodecBase()
                codec.SIZE = 0
                codec.IS_PATH = False
            else:
                codec = codec_by_name(proto.codec)
        except (ImportError, exceptions.ProtocolNotFoundError) as exc:
            raise exceptions.StringParseError("Unknown Protocol", string, element) from exc
```

### 2. Test Changes

#### tests/test_multiaddr.py
The test suite for the Multiaddr class has been enhanced with the following changes:

1. Sequence Behavior Tests
   - Added new test cases for sequence behavior
   - Enhanced sequence operation testing
   - Improved sequence validation
   - Added sequence edge case testing

2. Protocol Handling Tests
   - Enhanced test coverage for protocol handling
   - Improved protocol operation testing
   - Added protocol validation testing
   - Enhanced protocol edge case testing

3. Error Handling Tests
   - Improved test cases for error handling
   - Enhanced error message testing
   - Added error recovery testing
   - Improved error edge case testing

4. Negative Indexing Tests
   - Added tests for negative indexing
   - Enhanced index validation testing
   - Improved index edge case testing
   - Added index operation testing

5. Protocol Validation Tests
   - Enhanced test cases for protocol validation
   - Improved validation rule testing
   - Added validation edge case testing
   - Enhanced validation operation testing

6. Path-based Protocol Tests
   - Added tests for path-based protocols
   - Enhanced path validation testing
   - Improved path operation testing
   - Added path edge case testing

Key changes:
```python
def test_sequence_behavior():
    ma = Multiaddr("/ip4/127.0.0.1/udp/1234")
    proto1 = protocol_with_name("ip4")
    proto2 = protocol_with_name("udp")
    value1 = "127.0.0.1"
    value2 = "1234"
    item1 = (proto1, value1)
    item2 = (proto2, value2)

    # Test positive indices
    for idx, (proto, value, item) in enumerate(zip([proto1, proto2], [value1, value2], [item1, item2])):
        assert proto in ma
        assert value in ma.values()
        assert item in ma.items()
```

#### tests/test_protocols.py
The test suite for the protocols module has been enhanced with the following changes:

1. Protocol Registry Tests
   - Enhanced test cases for protocol registry
   - Improved registry operation testing
   - Added registry validation testing
   - Enhanced registry edge case testing

2. Protocol Alias Tests
   - Improved test coverage for protocol aliases
   - Enhanced alias operation testing
   - Added alias validation testing
   - Improved alias edge case testing

3. Protocol Locking Tests
   - Added tests for protocol locking
   - Enhanced lock operation testing
   - Improved lock validation testing
   - Added lock edge case testing

4. Error Handling Tests
   - Enhanced test cases for error handling
   - Improved error message testing
   - Added error recovery testing
   - Enhanced error edge case testing

5. Protocol Validation Tests
   - Added tests for protocol validation
   - Enhanced validation rule testing
   - Improved validation operation testing
   - Added validation edge case testing

Key changes:
```python
def test_add_protocol_alias():
    registry = protocols.REGISTRY.copy(unlock=True)
    tcp_proto = protocols.protocol_with_name("tcp")
    registry.add_alias_name("tcp", "abcd")
    registry.add_alias_code(tcp_proto, 123456)

    with pytest.raises(exceptions.ProtocolExistsError):
        registry.add_alias_name("tcp", "abcd")
    with pytest.raises(exceptions.ProtocolExistsError):
        registry.add_alias_code(tcp_proto, 123456)
```

#### tests/test_transforms.py
The test suite for the transforms module has been enhanced with the following changes:

1. CID Conversion Tests
   - Added support for CIDv0 to CIDv1 conversion
   - Enhanced conversion testing
   - Improved conversion validation
   - Added conversion edge case testing

2. CID Handling Tests
   - Enhanced test cases for CID handling
   - Improved CID operation testing
   - Added CID validation testing
   - Enhanced CID edge case testing

3. CID Conversion Behavior Tests
   - Improved validation of CID conversion behavior
   - Enhanced conversion rule testing
   - Added conversion operation testing
   - Improved conversion edge case testing

4. Path-based Protocol Tests
   - Added tests for path-based protocols
   - Enhanced path validation testing
   - Improved path operation testing
   - Added path edge case testing

Key changes:
```python
@pytest.mark.parametrize("proto, string, expected", [
    (REGISTRY.find('p2p'),  # This one gets autoconverted to CIDv1
     "12D3KooWPA6ax6t3jqTyGq73Zm1RmwppYqxaXzrtarfcTWGp5Wzx",
     b"\x01\x72\x00\x24\x08\x01\x12\x20\xc6\x35\xed\x47\x2d\x58\xed\xd8\xb9\xee\x46\x41"
     b"\xec\x75\x57\xdc\x80\x6b\x48\x86\xe5\x68\x49\x1a\xb0\x0b\xff\x12\x69\xa7\xbe\x65"),
    (REGISTRY.find('ip6'),  # Others do not
     "12D3KooWPA6ax6t3jqTyGq73Zm1RmwppYqxaXzrtarfcTWGp5Wzx",
     b"\x00\x24\x08\x01\x12\x20\xc6\x35\xed\x47\x2d\x58\xed\xd8\xb9\xee\x46\x41\xec\x75"
     b"\x57\xdc\x80\x6b\x48\x86\xe5\x68\x49\x1a\xb0\x0b\xff\x12\x69\xa7\xbe\x65"),
])
def test_cid_autoconvert_to_bytes(proto, string, expected):
    assert codec_by_name("cid").to_bytes(proto, string) == expected
```

## Summary

The changes reflect a focus on:
1. Improving type safety throughout the codebase
   - Enhanced type hints and annotations
   - Improved type checking and validation
   - Added support for generic types
   - Enhanced type safety features

2. Enhancing protocol handling and validation
   - Improved protocol registration and lookup
   - Enhanced protocol validation rules
   - Added support for protocol operations
   - Improved protocol performance

3. Adding support for newer CID formats
   - Added CIDv1 support
   - Enhanced CID conversion
   - Improved CID validation
   - Added CID operations

4. Improving error handling and reporting
   - Enhanced error messages
   - Added specific exception types
   - Improved error recovery
   - Enhanced error validation

5. Making the codebase more robust and maintainable
   - Improved code organization
   - Enhanced code quality
   - Added code documentation
   - Improved code performance

6. Enhancing string parsing and validation
   - Improved string format validation
   - Enhanced parsing error handling
   - Added support for new formats
   - Improved parsing performance

7. Improving protocol registry handling
   - Enhanced registry operations
   - Improved registry validation
   - Added registry features
   - Enhanced registry performance

8. Increasing test coverage and validation
   - Added new test cases
   - Enhanced existing tests
   - Improved test organization
   - Added test documentation

These updates ensure:
- Better compatibility with Python's type system
- More reliable protocol handling
- Enhanced support for modern address formats
- Improved error handling and debugging
- Better maintainability and code quality
- More robust string parsing and validation
- Enhanced protocol registry functionality
- Comprehensive test coverage

The changes maintain backward compatibility while adding new features and improving existing functionality. The test suite has been significantly enhanced to provide better coverage and validation of the library's features.

**Comment by seetadev** [comment]
@arcinston , @acul71 : We are heading towards getting ownership of multiformats/pymultiaddrs.

As shared by @pacrob, lets try and ensure all type errors have been resolved within Py-libp2p locally first related to pymultiaddrs.

This would also help other initiatives within py-libp2p.

**Comment by acul71** [comment]
> @arcinston , @acul71 : We are heading towards getting ownership of multiformats/pymultiaddrs.
> 
> As shared by @pacrob, lets try and ensure all type errors have been resolved within Py-libp2p locally first related to pymultiaddrs.
> 
> This would also help other initiatives within py-libp2p.

I'll work on this May 31 (Saturday night) and Sunday.
About `multiformats/pymultiaddrs` I've fixed the type errors here:
https://github.com/acul71/py-multiaddr-fork/tree/fix/typecheck

I've modified py-multiaddr-fork setup with this 
[fix: quick and dirty adding typecheck](https://github.com/multiformats/py-multiaddr/commit/8562f14cc2031a49b2a6bf4c00115482a99b3c01) @arcinston  can you check if it's allright and suggest corrections?

I don't know exactly how to incorporate the py-multiaddr-fork in our py-libp2p.
I'm unsure how to proceed. So I've asked AI how to do this.
Can you tell if this is the right path?
Thanks.

# Using Modified py-multiaddr with py-libp2p

This guide describes how to use a modified version of py-multiaddr with py-libp2p while the main py-multiaddr repository is archived.

## 1. Setup Local Development Environment

```bash
# 1. Clone your py-multiaddr fork if you haven't already
git clone https://github.com/acul71/py-multiaddr-fork.git
cd py-multiaddr-fork
git checkout fix/typecheck

# 2. Create and activate virtual environment
python3 -m venv ./venv
. venv/bin/activate  # On Windows: venv\Scripts\activate

# 3. Install development dependencies
pip install -r requirements_dev.txt
pip install -e .

# 4. Setup pre-commit hooks
pre-commit install
```

## 2. Development Tools

The fork includes several development tools:

```bash
# Run type checking
make typecheck

# Run linting and auto-fix
make fix

# Run tests
make test

# Run tests on specific Python versions
make test-all
```

## 3. Modify py-libp2p's Dependencies

Since py-libp2p has a hard dependency on multiaddr in its `pyproject.toml`, you have two options:

### a) Temporary Solution (for development)

```bash
# Create and activate a new virtual environment
cd py-libp2p
python3 -m venv ./venv
. venv/bin/activate  # On Windows: venv\Scripts\activate

# Install your modified py-multiaddr first
cd ../py-multiaddr-fork
pip install -e .

# Install py-libp2p with development dependencies
cd ../py-libp2p
python3 -m pip install -e ".[dev]"
pre-commit install
```

### b) Permanent Solution (for contribution)

Fork py-libp2p and modify the dependency in `pyproject.toml` to point to your fork:

```toml
dependencies = [
    # ... other dependencies ...
    "multiaddr @ git+https://github.com/acul71/py-multiaddr-fork.git@fix/typecheck",
    # ... other dependencies ...
]
```

Then follow the standard py-libp2p setup:
```bash
cd py-libp2p
python3 -m venv ./venv
. venv/bin/activate
python3 -m pip install -e ".[dev]"
pre-commit install
```

## 4. Testing Your Changes

```bash
# Run py-libp2p's test suite to ensure compatibility
cd path/to/py-libp2p
pytest tests/

# Run specific tests that use multiaddr functionality
pytest tests/core/transport/
pytest tests/utils/interop/
```

## 5. When the Main Repository Reopens

- Create a pull request to the main py-multiaddr repository
- Once merged, you can update py-libp2p to use the official version again

## 6. Development Workflow

1. Make changes to your py-multiaddr fork
2. Run type checking: `make typecheck`
3. Run linting: `make fix`
4. Run tests: `make test`
5. Install the updated version: `pip install -e .` in the py-multiaddr directory
6. Test the changes in py-libp2p
7. Repeat as needed

## 7. Keeping Track of Changes

- Document your changes in your fork's README
- Keep your fork's `fix/typecheck` branch up to date with the main repository
- Consider creating a changelog to track your modifications

## 8. Best Practices

- Use version control to track your changes
- Write tests for your modifications
- Document any breaking changes
- Keep your fork's dependencies up to date
- Run type checking and linting before committing changes

## 9. Sharing with Other Developers

Other py-libp2p developers can use your modified version by:

1. **Using Your Fork**:
   - Replace the repository URL in their `pyproject.toml` with:
   ```toml
   "multiaddr @ git+https://github.com/acul71/py-multiaddr-fork.git@fix/typecheck"
   ```

2. **Following This Guide**:
   - Share this `MULTIADDR_SETUP.md` file with them
   - They can follow the same setup procedure

3. **Contributing Back**:
   - Other developers can submit issues or pull requests to your fork
   - This helps maintain the solution until the main repository reopens

4. **Staying Updated**:
   - Watch your fork's repository for updates
   - Pull the latest changes when available

## Note

This is a temporary solution until the main py-multiaddr repository is reopened. Once it's reopened, you should submit your changes as a pull request to the main repository.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -32,11 +32,15 @@ repos:
...
         always_run: true
         pass_filenames: false
-  - repo: https://github.com/northisup/pyright-pretty
-    rev: v0.1.0
+  - repo: local
     hooks:
-      - id: pyright-pretty
-        args: [--show-summary]
+      - id: pyrefly-local
+        name: run prefly typecheck locally
```
Comment (acul71) [inline .pre-commit-config.yaml]: Should this be 
name: run pyrefly typecheck locally

**Review by acul71 — APPROVED** [review]

File: .pre-commit-config.yaml (hunk 1 above)
Comment (arcinston) [reply .pre-commit-config.yaml]: yess , thanks for pointing out

**Comment by seetadev** [comment]
@arcinston : Great progress. Could you please share a latest status report on the remaining type errors. 

@lla-dane , @varun-r-mallya , @sukhman-sukh , @acul71 : Thank you for your contribution. Appreciate it.

I had a discussion today with the team. Wish if we could try and arrive at a good conclusion on the issues by the end of this week, if possible. If not, absolutely fine. Great work indeed.

File: pyproject.toml
Comment (pacrob) [inline pyproject.toml:1]: It might just be personal preference, but can ruff and pyrefly config be in here instead of adding separate config files?

Comment (arcinston) [reply pyproject.toml:1]: yup this can be done 

and I agree in this case

kept it seperate during dev for an easier review process and (testing stuff out)

but will move it to the pyproject.toml 💯

**Comment by pacrob** [comment]
The `.project-template` folder and contents can be deleted. This repo was originally cloned from https://github.com/ethereum/ethereum-python-project-template , but this PR makes so many changes that it's no longer relevant.

File: examples/doc-examples/example_encryption_insecure.py [hunk 2]
```diff
@@ -26,7 +26,6 @@ async def main():
         local_key_pair=key_pair,
         # secure_bytes_provider: Optional function to generate secure random bytes
         # (defaults to secrets.token_bytes)
-        secure_bytes_provider=None,  # Use default implementation
     )
```
Comment (pacrob) [inline examples/doc-examples/example_encryption_insecure.py]: Why is `secure_bytes_provider` being removed? And if removing, remove the comment too.

File: examples/doc-examples/example_encryption_secio.py [hunk 3]
```diff
@@ -24,7 +26,6 @@ async def main():
         local_key_pair=key_pair,
         # secure_bytes_provider: Optional function to generate secure random bytes
         # (defaults to secrets.token_bytes)
-        secure_bytes_provider=None,  # Use default implementation
     )
```
Comment (pacrob) [inline examples/doc-examples/example_encryption_secio.py]: same question as in `example_encryption_insecure.py`

File: examples/doc-examples/example_encryption_insecure.py (hunk 2 above)
Comment (arcinston) [reply examples/doc-examples/example_encryption_insecure.py]: will do

**Comment by arcinston** [comment]
main has been merged (conflicts resolved) & all type errors have now been fixed , 
can run the CI now

File: libp2p/protocol_muxer/multiselect.py [hunk 4]
```diff
@@ -41,9 +48,10 @@ def add_handler(self, protocol: TProtocol, handler: StreamHandlerFn) -> None:
         """
         self.handlers[protocol] = handler
 
+    # FIXME: Make TProtocol Optional[TProtocol] to keep types consistent
```
Comment (pacrob) [inline libp2p/protocol_muxer/multiselect.py:48]: Still to be fixed in this PR?

File: libp2p/pubsub/pubsub.py [hunk 5]
```diff
@@ -120,7 +123,7 @@ class Pubsub(Service, IPubsub):
 
     # Indicate if we should enforce signature verification
     strict_signing: bool
-    sign_key: PrivateKey
+    sign_key: PrivateKey | None
```
Comment (pacrob) [inline libp2p/pubsub/pubsub.py:126]: Use `Optional` instead of `|`

File: libp2p/pubsub/pubsub.py [hunk 6]
```diff
@@ -132,7 +135,7 @@ def __init__(
         self,
         host: IHost,
         router: IPubsubRouter,
-        cache_size: int = None,
+        cache_size: int | None = None,
```
Comment (pacrob) [inline libp2p/pubsub/pubsub.py:138]: Use `Optional` instead of `|`

File: libp2p/security/secure_session.py [hunk 7]
```diff
@@ -85,6 +85,10 @@ async def read(self, n: int = None) -> bytes:
 
         msg = await self.conn.read_msg()
 
+        # FIXME: I am not very sure of this fix
```
Comment (pacrob) [inline libp2p/security/secure_session.py]: To be fixed in this PR?

File: libp2p/stream_muxer/mplex/mplex.py [hunk 8]
```diff
@@ -138,7 +138,7 @@ async def _initialize_stream(self, stream_id: StreamID, name: str) -> MplexStrea
         send_channel, receive_channel = trio.open_memory_channel[bytes](
             MPLEX_MESSAGE_CHANNEL_SIZE
         )
-        stream = MplexStream(name, stream_id, self, receive_channel)
+        stream = MplexStream(name, stream_id, self, receive_channel)  # type: ignore
```
Comment (pacrob) [inline libp2p/stream_muxer/mplex/mplex.py]: If `type: ignore` is appropriate, add a note explaining why

File: libp2p/tools/async_service/trio_service.py [hunk 9]
```diff
@@ -309,7 +320,7 @@ def run_task(
         async_fn: Callable[..., Awaitable[Any]],
         *args: Any,
         daemon: bool = False,
-        name: str = None,
+        name: str | None = None,
```
Comment (pacrob) [inline libp2p/tools/async_service/trio_service.py:323]: Use `Optional` instead of `|`

File: libp2p/tools/async_service/trio_service.py [hunk 10]
```diff
@@ -322,7 +333,7 @@ def run_task(
         self._common_run_task(task)
 
     def run_child_service(
-        self, service: ServiceAPI, daemon: bool = False, name: str = None
+        self, service: ServiceAPI, daemon: bool = False, name: str | None = None
```
Comment (pacrob) [inline libp2p/tools/async_service/trio_service.py:336]: Use `Optional` instead of `|`

File: libp2p/tools/async_service/trio_service.py [hunk 11]
```diff
@@ -416,7 +427,12 @@ async def inner(self: ServiceAPI, *args: Any, **kwargs: Any) -> Any:
         async with trio.open_nursery() as nursery:
             # mypy's type hints for start_soon break with this invocation.
             nursery.start_soon(
-                _wait_api_fn, self, func, args, kwargs, send_channel  # type: ignore
+                _wait_api_fn,  # type: ignore
```
Comment (pacrob) [inline libp2p/tools/async_service/trio_service.py:430]: This is a preexisting `type: ignore`, but can it be resolved in this PR? Fine if not.

File: tests/core/stream_muxer/test_yamux.py [hunk 12]
```diff
@@ -215,8 +226,14 @@ async def test_yamux_stream_reset(yamux_pair):
     server_stream = await server_yamux.accept_stream()
     await client_stream.reset()
     # After reset, reading should raise MuxedStreamReset or MuxedStreamEOF
-    with pytest.raises((MuxedStreamEOF, MuxedStreamError)):
+    # with pytest.raises((MuxedStreamEOF, MuxedStreamError)):
```
Comment (pacrob) [inline tests/core/stream_muxer/test_yamux.py:229]: clean up?

File: tests/core/stream_muxer/test_yamux.py [hunk 13]
```diff
@@ -215,8 +226,14 @@ async def test_yamux_stream_reset(yamux_pair):
...
+    #     await server_stream.read()
```
Comment (pacrob) [inline tests/core/stream_muxer/test_yamux.py]: clean up?

File: tests/core/transport/test_tcp.py [hunk 14]
```diff
@@ -36,7 +36,7 @@ async def handler(tcp_stream):
 @pytest.mark.trio
 async def test_tcp_dial(nursery):
     transport = TCP()
-    raw_conn_other_side = None
+    raw_conn_other_side: RawConnection | None = None
```
Comment (pacrob) [inline tests/core/transport/test_tcp.py:39]: Use `Optional` instead of `|`

File: tests/core/network/test_notify.py [hunk 15]
```diff
@@ -40,169 +37,94 @@ class Event(enum.Enum):
...
+        #     events_1_0, Event.ClosedStream, 1.0
+        # )  # Not implemented
+        assert await wait_for_event(events_1_0, Event.Disconnected, 1.0)
+
+        assert await wait_for_event(events_1_1, Event.Connected, 1.0)
+        assert await wait_for_event(events_1_1, Event.OpenedStream, 1.0)
+        # assert await wait_for_event(
+        #     events_1_1, Event.ClosedStream, 1.0
+        # )  # Not implemented
+        assert await wait_for_event(events_1_1, Event.Disconnected, 1.0)
```
Comment (pacrob) [inline tests/core/network/test_notify.py:130]: Should these commented `assert`s still be here?

**Comment by pacrob** [comment]
OK! I've looked through everything. Once comments are addressed, I'm good to merge.

File: tests/core/network/test_notify.py (hunk 15 above)
Comment (arcinston) [reply tests/core/network/test_notify.py:130]: should be commented for now

afaik @seetadev did inform me , notify is being re-worked and implemented currently

**Comment by arcinston** [comment]
thank you @pacrob for looking closely at the PR
fixing these right away

**Comment by arcinston** [comment]
@pacrob the reason at some places " | None "  is used over Optional is due to the precommit hook
pyupgrade for some files to "| None"

the reason for this is after python 3.10 -> Type | None was preffered over Optional

source :https://stackoverflow.com/questions/69440494/python-3-10-optionaltype-or-type-none

changing the version for pyupgrade to 3.10 converts all optional types to Type | None

two choices here

1. accept the newer syntax of Type | None over Optional[Type] as preffered by python and update pyupgrade to 3.10 plus
2. remove pyypgrade from pre-commit and continue to use Optional 

personally would prefer continue using pyupgrade hence can shift to using Type | None over Optional , but no hard opinions either way

**Comment by arcinston** [comment]
ive fixed all the reviewed changes except the above pending decision / discussion to be had

**Comment by seetadev** [comment]
> @pacrob the reason at some places " | None " is used over Optional is due to the precommit hook pyupgrade for some files to "| None"
> 
> the reason for this is after python 3.10 -> Type | None was preffered over Optional
> 
> source :https://stackoverflow.com/questions/69440494/python-3-10-optionaltype-or-type-none
> 
> changing the version for pyupgrade to 3.10 converts all optional types to Type | None
> 
> two choices here
> 
> 1. accept the newer syntax of Type | None over Optional[Type] as preffered by python and update pyupgrade to 3.10 plus
> 2. remove pyypgrade from pre-commit and continue to use Optional
> 
> personally would prefer continue using pyupgrade hence can shift to using Type | None over Optional , but no hard opinions either way

@arcinston : Thanks for the detailed context and feedback, Arush — appreciate it.

You're absolutely right that post-Python 3.10, the `Type | None` syntax is not only preferred but also more readable in many cases. Thanks for linking the StackOverflow thread as well — it provides solid reasoning. I'm aligned with your thought process here.

We are also in favor of leaning into the modern syntax (`Type | None`) and updating `pyupgrade` accordingly. It's clean, aligns with the latest Python standards, and avoids extra imports from `typing`. Since we're already on 3.10+ across the board (or aiming to be), this feels like a natural evolution for the codebase.

That said, we are okay with consistency being the priority too. If our contributors prefer sticking with `Optional[Type]` for compatibility or style reasons, I wouldn’t strongly oppose. But overall, +1 from me on going ahead with `Type | None` and keeping `pyupgrade` in the workflow.

Let’s just make sure we document the decision clearly in the contributing guide or a dev note so it’s easy for future contributors to follow. We can add these pointers at the discussion page opened for developer on-boarding to the new stack.

**Comment by arcinston** [comment]
thanks for the comment @seetadev 
have made the respective changes

**Comment by pacrob** [comment]
@arcinston @seetadev I wasn't planning to drop python 3.9 support until it is officially deprecated later this year. That said, I'm fine with doing so now, and I do personally prefer the `|` syntax.

In order to fully drop py39, it should be removed from CI and `classifiers` in `pyproject.toml`.

**Comment by seetadev** [comment]
@pacrob : Thank you for the thoughtful inputs and feedback. Appreciate it. 

Indeed, it's fair to wait until Python 3.9 is officially deprecated later this year, especially if we want to keep broader compatibility for a bit longer. 

That said, since we're fine with dropping it now and also prefer the `|` syntax, I’d fully support moving forward with Python 3.10+ as our baseline.

The `|` syntax does offer clearer readability and aligns well with the modern Python ecosystem — especially now that most tooling, libraries, and contributors are shifting towards 3.10+ as a standard. From a maintenance and contributor experience perspective, reducing the legacy syntax and aligning with what `pyupgrade` recommends also helps streamline the codebase.

@arcinston : If we do this, we should clean it up properly — removing Python 3.9 from the CI matrix and updating the `classifiers` in `pyproject.toml` to reflect the change. I’d also suggest a quick note in the changelog or `CONTRIBUTING.md` to highlight the version bump for anyone contributing downstream.

**Comment by seetadev** [comment]
@pacrob Thank you so much for the detailed and thoughtful feedback — it’s been incredibly insightful. This PR has surfaced a number of valuable learnings around software engineering best practices in Python projects. Honestly, there's enough here that all the contributors in this PR could co-author a paper for the Python Journal someday :)

@arcinston Great progress. Also, appreciate your efforts to document the developer onboarding guide for the new stack in the discussion thread — that’s going to be super helpful for everyone joining in.

Also, a big thank you to @lla-dane, @varun-r-mallya, @guha-rahul, @sukhman-sukh, @acul71, @Aldorax, @0-th for your great contributions and wonderful support throughout this process. 🙌

**Comment by arcinston** [comment]
@seetadev last commit cleans up the forementioned changes

**Comment by seetadev** [comment]
> @seetadev last commit cleans up the forementioned changes

@arcinston : Awesome, great to hear.

@pacrob : Hi Paul. Wish to share this PR is ready for a final review + merge.

File: libp2p/tools/async_service/trio_service.py (hunk 11 above)
Comment (pacrob) [reply libp2p/tools/async_service/trio_service.py:430]: Nevermind, `async_service` is being removed soon, no need to spend time on this.

**Comment by pacrob** [comment]
Looking good! One `FIXME` question outstanding in `libp2p/protocol_muxer/multiselect.py`. Otherwise, as @seetadev mentions, please add newsfragments to make the changes visible in the release notes.

2 newsfragments should be fine -  a `618.internal.rst` to describe added typing tools and related code updates, and a `618.removal.rst` for the removal of python 3.9 support.
